packages/core/core/src/services/document-service/transform/utils.ts

```diff
@@ -1,9 +1,72 @@
-const switchIdForDocumentId = (output: Record<string, any>) => {
+import { Common } from '@strapi/types';
+import { traverse } from '@strapi/utils';
+
+import { IdMap } from './id-map';
+
+export const switchIdForDocumentId = (output: Record<string, any>) => {
   // Mutating for performance reasons
   const documentId = output.documentId;
   delete output.documentId;
   output.id = documentId;
   return output;
 };
 
-export { switchIdForDocumentId };
+export const transformFieldsOrSort = (input: string[] | string): string[] | string => {
+  if (Array.isArray(input)) {
+    return input.map((item) => (item === 'id' ? 'documentId' : item));
+  }
+  return input.replace('id', 'documentId');
+};
+
+// TODO what if UID is undefined? throw?
+export const transformFilters = async (
+  idMap: IdMap,
+  data: Record<string, any>,
+  opts: {
+    uid: Common.UID.Schema;
+    locale?: string | null;
+  }
+) => {
+  return traverse.traverseQueryFilters(
+    async ({ attribute, key, value, path, schema }, { set, remove }) => {
+      if (!attribute) {
+        return;
+      }
+
+      // In order to correctly ignore nested keys that are not relations
+      // we keep track of the parent key within the path
+      // Ensuring that the parentKey is a relation of the current schema
+      // or the current schema itself
+      let parentKey: string | undefined;
+      if (typeof path.raw === 'string') {
+        const keys = path.raw.split('.');
+        parentKey = keys?.[keys.length - 2];
+
+        // Check if parentKey is an array and remove the indexing to get the real parentKey
+        if (parentKey && parentKey.endsWith(']')) {
+          parentKey = parentKey.slice(0, parentKey.lastIndexOf('['));
+        }
+      }
```

Comment:
@Marc-Roig, could I get your opinion on this? It was in order to manage situations such as:

As I understand it we only want to convert id -> documentId when we are wthin a relational attribute?

```
    it('should handle complex nested structures, ignoring nested non relational keys', async () => {
      const input = {
        relatedProducts: {
          categories: [{ id: 'complex1' }, { someKey: { id: 'complex2' } }],
        },
      };
      const expected = {
        relatedProducts: {
          categories: [{ documentId: 'complex1' }, { someKey: { id: 'complex2' } }],
        },
      };

      expect(await transformFilters(idMap, input, { uid: PRODUCT_UID })).toEqual(expected);
    });
```

Reply:
yes 🤔 Not really convinced we should check the strings like that, filters can be expressed in many ways and we might be forgetting about edge cases.
Want to play for a bit with this to be more familiar with the traversers 🤔

Reply:
Just to be sure, is this nested keys format something we need to cover at all?

Reply:
what if instead we:
- Traverse filters
- If land on a relational field
- Then check if it contains an id field
- If it does,  then set the value to document id

I believe we would not need to check for string patterns like that !